File: Baekjoon/1932_정수 삼각형/B_1932_3.py
import sys


n = int(sys.stdin.readline())
triangle = [list(map(int, sys.stdin.readline().split())) for _ in range(n)]
dp = [[0] * n for _ in range(n)]
dp[0][0] = triangle[0][0]

for i in range(1, n):
    t = triangle[i]
    t_end = len(t) - 1
    for j in range(len(t)):
        if j == 0:
            dp[i][j] = dp[i - 1][j]
        elif j == t_end:
            dp[i][j] = dp[i - 1][j - 1]
        else:
            dp[i][j] = max(dp[i - 1][j - 1], dp[i - 1][j])
        dp[i][j] += triangle[i][j]

print(max(dp[n - 1]))

File: Baekjoon/1932_정수 삼각형/B_1932_3.js
const setDp = (arr) => {
  for (let i = 1; i <= n; i++) {
    arr.push(Array.from(Array(i).fill(0)));
  }

  return arr;
}

const getMaxValue = () => {
  for (let i = 1; i < n; i++) {
    for (let j = 0; j < dp[i].length; j++) {
      if (j === 0) {
        dp[i][j] = dp[i - 1][j] + triangle[i][j];
        continue;
      }

      if (j === i) {
        dp[i][j] = dp[i - 1][j - 1] + triangle[i][j];
        continue;
      }

      dp[i][j] = Math.max(dp[i - 1][j - 1], dp[i - 1][j]) + triangle[i][j];
    }
  }

  return dp;
}


const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

// const input = [
//   '5',
//   '7',
//   '3 8',
//   '8 1 0',
//   '2 7 4 4',
//   '4 5 2 6 5'
// ];
// => 30

const n = parseInt(input[0]);
input.shift();

const triangle = input.map(v => v.split(' ').map(v => parseInt(v)));
const dp = setDp([]);
dp[0][0] = triangle[0][0]

getMaxValue();
console.log(Math.max(...dp[n - 1]));
